content/exchange/artifacts/Linux.Debian.GPGKeys.yaml

```diff
@@ -0,0 +1,154 @@
+name: Linux.Debian.GPGKeys
+description: |
+  Extract keys, fingerprints and identities from GPG keys.
+
+  This artifact runs the tool "gpg" (must be installed on the system) on the
+  files found matching the globs in KeyringFiles. The files need not be keyrings.
+
+  Every entry consists of a public or secret key, optional subkeys and optional
+  identities. This artifact may be useful in other artifacts to inspect GPG
+  files or GPG data in order to correlate keys by their IDs, or look at connected
+  user IDs.
+
+  This artifact doesn't provide any information about whether a key is
+  "trustworthy".
+
+  Note that some keyring files contain a lot of subkeys and identities.
+
+  The following columns are returned by this artifact:
+
+    - OSPath: Path to the key file
+    - KeyInfo: dict with the following entries:
+      - Type: pub|sub
+      - ID
+      - Fingerprint
+      - Algorithm
+      - Validity
+      - Created
+      - Expiry
+    - SubKeys: array of dicts with the same structure as KeyInfo
+    - UserIDs: array of strings (name and e-mail)
+
+reference:
+  - https://manpages.debian.org/bookworm/apt/apt-key.8.en.html
+  - https://github.com/CSNW/gnupg/blob/master/doc/DETAILS
+  - https://www.mailpile.is/blog/2014-10-07_Some_Thoughts_on_GnuPG.html
+  - https://www.ietf.org/rfc/rfc4880.txt
+
+export: |
+        /* Extract machine-"readable" data from the GPG keys found in the file.
+           The format is documented in the reference above. However, as the blog
+           post mentions, detailed knowledge about GPG is needed in order to
+           decipher the output. See ParseKeyInfo_(). */
+        LET InspectGPGFile(filename) = SELECT Stdout AS Info
+            FROM execve(argv=['gpg', '--with-colons', filename])
+
+        /* Pipe data to the same command as in InspectGPGFile(): */
+        LET InspectGPGData(data) = SELECT Stdout AS Info
+            FROM execve(argv=['sh', '-c', format(format='echo "%v"|gpg --with-colons', args=[data])])
```

Comment:
This can open us up to command line injection. It might be safer to use a temp file here?

Reply:
Absolutely, I'll fix that.